### PR TITLE
Fixup subplans

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
Fixup subplans referring to same plan_id: Before parallelization on nodes in cdbparallelize if there are any subplan nodes in the plan which refer to the same plan_id, parallelization step breaks as a node must be processed only once by it.
This patch fixes the issue by generating a new subplan node in glob subplans, and updating the plan_id of the subplan to refer to the newly created node.

Skip subplan nodes in equivalence lists: Skip duplicating subplans clauses as multiple subplan node referring to same plan_id of subplans breaks cdbparallize. cdbparallize expects to apply motion only to a node once, so if two subplan nodes refers to the same plan_id the assertion breaks.

### CODE REVIEW

**Comment by hlinnaka** [comment]
This still fails with this patch:

```
CREATE TABLE foo (a integer, b integer)  PARTITION BY RANGE(b)
    (PARTITION sub_one START (1) END (10),
     PARTITION sub_two START (11) END (22));
CREATE TABLE bar (c integer, d character varying(10));
insert into foo values (9,9);
insert into bar values (9,9);
create table baz (i int4);

-- With dynamic partition pruning enabled you run into the bug discussed at PR #2444
set gp_dynamic_partition_pruning=off;

explain select bar.c from bar, foo where foo.b = (select max(i) from baz where bar.c = 9) and foo.b = bar.d::int4;
FATAL:  Unexpected internal error (nodeMotion.c:934)
DETAIL:  FailedAssertion("!(recvSlice->gangSize == 1)", File: "nodeMotion.c", Line: 934)
```

**Comment by hlinnaka** [comment]
The first commit, 1d4e26ee2696d4193de891b2c9a40782c6a47db9, is enough to fix the test query that's added to the regression suite. Do you have a test query that needs the second commit?

**Comment by hsyuan** [comment]
The query in issue #2440 is intended to test the 2nd commit, I guess.

**Comment by bhuvnesh2703** [comment]
@hsyuan @hlinnaka. Updated the PR.
Before, i thought the query listed in #2440 should require the 2nd commit available in this pull request. But looks like not, after the fix to not consider subplans in partition selection key. So for now have removed the code to duplicate subplans if existing. And have included the fix as suggested in the PR #2444. 
pls review 
Will open up a new issue for the failure mentioned above by @hlinnaka `FailedAssertion("!(recvSlice->gangSize == 1)", File: "nodeMotion.c"`

**Comment by bhuvnesh2703** [comment]
@hlinnaka @hsyuan incase the current commit `868f92f772e2eb9b562dda78e2acc4c06fe65d4f` in this PR needs to be applied only after the issue described by Heikki is fixed, let me know.

**Comment by bhuvnesh2703** [comment]
@hlinnaka @hsyuan created another issue: https://github.com/greenplum-db/gpdb/issues/2454
Will raise a PR for that after this PR goes in as i based it on this commit.
The below fix avoids issue:
```
diff --git a/src/backend/optimizer/plan/createplan.c b/src/backend/optimizer/plan/createplan.c
index a3166ed108..288026ed39 100644
--- a/src/backend/optimizer/plan/createplan.c
+++ b/src/backend/optimizer/plan/createplan.c
@@ -4519,7 +4519,8 @@ make_sort_from_pathkeys(PlannerInfo *root, Plan *lefttree, List *pathkeys,
                                        List       *exprvars;
                                        ListCell   *k;

-                                       if (em->em_is_const || em->em_is_child)
+                                       /* Skip SubPlans * in the sort keys */
+                                       if (em->em_is_const || em->em_is_child || contain_subplans((Node *) em->em_expr))
                                                continue;
                                        sortexpr = em->em_expr;
                                        exprvars = pull_var_clause((Node *) sortexpr, false);
```
Also another questions, do you think we can use IsA(node, SubPlan) instead of using contain_subplans. Do you suspect any cases where it might break?

**Comment by hsyuan** [comment]
We can't just use IsA(node,SubPlan). because it can be A+Subplan, or
function(SubPlan).....

~ ~ ~
Haisheng Yuan

**Comment by bhuvnesh2703** [comment]
@hsyuan thanks

File: src/backend/optimizer/path/equivclass.c [hunk 1]
```diff
@@ -690,7 +690,14 @@ generate_base_implied_equalities_no_const(PlannerInfo *root,
...
 											 cur_em->em_datatype);
-			if (!OidIsValid(eq_op))
+			/*
+			 * Skip duplicating subplans clauses as multiple subplan node referring
+			 * to the same plan node fails the assertion made by the code which adds
+			 * motion to the plan
+			 */
+			if (contain_subplans((Node*) prev_em->em_expr)
+				|| contain_subplans((Node*) cur_em->em_expr)
+				|| !OidIsValid(eq_op))
```
Comment (d) [inline src/backend/optimizer/path/equivclass.c]: If an equivalence class contains one expression that has a SubPlan, this code here actually bails the entire routine. What about the other pairs in the equivalence class, don't we want to generate those derived quals?

Omer and Jesse

**Comment by bhuvnesh2703** [comment]
@hsyuan Pushed the fix for #2454 in this PR

File: src/backend/optimizer/path/equivclass.c (hunk 1 above)
Comment (vraghavan78) [inline src/backend/optimizer/path/equivclass.c]: We should only not derive the quals with the subplan.

**Comment by bhuvnesh2703** [comment]
Applied feedback. pls review.
@hlinnaka the commit 3 query uses the code to duplicate subplans, queries are available in test case.

File: src/backend/cdb/cdbmutate.c [hunk 2]
```diff
@@ -3004,44 +3000,93 @@ subplan_walker(Node *node, SubPlanWalkerContext *context)
...
+/*
+ * Remove unused subplans from PlannerGlobal subplans
+ */
+void
+remove_unused_subplans(PlannerInfo *root, SubPlanWalkerContext *context)
+{
 
-	for (i = 0; i < num_subplans; i++)
+	ListCell *lc;
+	for (int i = 1; i <= list_length(root->glob->subplans); i++)
```
Comment (jpatel-pivotal) [inline src/backend/cdb/cdbmutate.c]: Why did we change the for loop to start at 1 vs. 0?

Comment (bhuvnesh2703) [inline src/backend/cdb/cdbmutate.c]: We insert the plan_id to the bitmapset in fixup_subplan_walker and it starts from 1. However, Earlier in subplan_walker we were inserting plan_id -1 (not the actual plan id), so were using the i = 0 which is the index of a subplan in root->glob->subplans. Now when we get the item `lc = list_nth_cell(root->glob->subplans, i-1);` we use -1 to get the index of the plan referred by plan_id.
Though, with this comment i realize i didn't remove the old subplan_walker, its not used now so will remove it.

**Comment by bhuvnesh2703** [comment]
any more comments? if not will go ahead and merge it by tomorrow.

**Pull request merged**